sympy/solvers/solvers.py

```diff
@@ -962,11 +962,16 @@ def _sympified_list(w):
         f[i] = f[i].replace(lambda w: isinstance(w, HyperbolicFunction) and \
             w.has_free(*symbols), lambda w: w.rewrite(exp))
 
-        # if we have a Matrix, we need to iterate over its elements again
         if f[i].is_Matrix:
-            bare_f = False
-            f.extend(list(f[i]))
-            f[i] = S.Zero
+            # Raise an error for matrices with symbolic shape to prevent infinite loops.
+            if any(isinstance(dim, Symbol) for dim in f[i].shape):
+                raise ValueError("Matrix with symbolic shape cannot be represented explicitly.")
+            else:
+                # If dimensions are explicit integers
+                # we need to iterate over the matrix elements.
+                bare_f = False
+                f.extend(list(f[i]))
```

Comment:
It would be better to call `as_explicit`. It is not sufficient to just check for Symbol because the shape can be an expression containing symbols.